### ISSUE TITLE
NMS for ExecuTorch at Kotlin

### ISSUE THREAD

**Issue by kejndan** [ultralytics/ultralytics#22805]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello. Can you please suggest which is the best practices for using non maximum suppression for executorch on kotlin? And is it possible to see a sample code for full postprocessing on kotlin somewhere?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @kejndan, thank you for your interest in Ultralytics and for the detailed question about ExecuTorch and Kotlin post-processing 🚀 This is an automated response to help you get started while an Ultralytics engineer reviews your issue and follows up with more specific guidance soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered. While the examples are primarily in Python, they also clarify the expected input/output formats and post-processing steps (including NMS) that you can mirror in Kotlin.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom deployment or integration ❓ Question (like your ExecuTorch + Kotlin scenario), please provide as much information as possible, including:
-Your model type and export format
-Example input shapes and expected outputs
-Any current Kotlin or ExecuTorch code snippets you are using
-Details on what specifically you are unsure about in the NMS or postprocessing pipeline

Also verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) if your question is related to model performance or output quality.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community 💬

## Upgrade

Before proceeding, please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8** to verify your issue or question is not already addressed by the latest changes:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Nice question, thanks for raising this.

On Android/Kotlin with ExecuTorch the usual pattern is to export YOLO11 with `model.export(format="executorch")`, run the `.pte` via the ExecuTorch Android APIs, and then implement postprocessing (decode + confidence filter + NMS) on the Kotlin side, mirroring the Python logic. The ExecuTorch model output layout matches the `prediction` tensor expected by our `non_max_suppression` function, so the most robust approach is to port the algorithm from [`ultralytics.utils.nms.non_max_suppression`](https://docs.ultralytics.com/reference/utils/nms/) and, if you need more detail, the supporting `TorchNMS` code there rather than writing NMS from scratch. For a conceptual overview of how NMS fits into the pipeline, you can also refer to the [Non-Maximum Suppression (NMS) Explained](https://www.ultralytics.com/glossary/non-maximum-suppression-nms) article.

We don’t currently ship a full end‑to‑end Kotlin postprocessing example (boxes/cls decoding + NMS), so you’ll need to adapt the Android snippet in the [ExecuTorch integration guide](https://docs.ultralytics.com/integrations/executorch/) by adding your own Kotlin NMS utility (e.g., `fun nms(boxes: FloatArray, scores: FloatArray, iouThresh: Float): IntArray`) that follows the same steps as the Python implementation.
